File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.util.*;

class Solution {
    public int solution(String numbers) {
        int answer = 0;

        Set<Integer> sample = new HashSet<>();
        for (int size = 1; size <= numbers.length(); size++) {
            generate("", size, sample, new boolean[numbers.length()], numbers);
        }

        for (int num : sample) {
            if (isPrime(num)) answer++;
        }
        return answer;
    }

    private void generate(String temp, int size, Set<Integer> sample, boolean[] visited, String numbers) {
        if (temp.length() >= size) {
            sample.add(Integer.parseInt(temp));
            return;
        }

        for (int i = 0; i < numbers.length(); i++) {
            if (visited[i]) continue;
            visited[i] = true;
            generate(temp + numbers.charAt(i), size, sample, visited, numbers);
            visited[i] = false;
        }
    }

    private boolean isPrime(int num) {
        if (num < 2) return false;

        for (int i = 2; i <= (int) Math.sqrt(num); i++) {
            if (num % i == 0) return false;
        }
        return true;
    }
}

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.cpp
#include <string>
#include <vector>
#include <set>
#include <iostream>
#include <cmath>
using namespace std;

set<int> nums;

void recursive(string num, string& numbers, vector<bool> visited)
{
    nums.insert(stoi(num));
    
    for(int i = 0; i < numbers.length(); ++i)
    {
        if (visited[i] == true)
            continue;
        
        visited[i] = true;
        recursive(num + numbers[i], numbers, visited);
        visited[i] = false;
    }
}

int solution(string numbers) {
    int answer = 0;
    
    recursive("0", numbers, vector<bool> (numbers.size(), false));
    
    for(auto n : nums)
    {
        bool flag = true;
        if (n <= 1)
            continue;
        if (n == 2)
        {
            ++answer;
            continue;
        }
        
        for(int i = 2; i <= sqrt(n) && flag; ++i)
        {
            if (n % i == 0)
            {
                flag = false;
            }
        }
        
        if (flag)
        {
            ++answer;
        }
    }
    
    return answer;
}